sys/net/network_layer/ng_ndp/ng_ndp.c

```diff
@@ -0,0 +1,1390 @@
+                             ng_ndp_opt_t *sl2a_opt)
+{
+    ng_ipv6_nc_t *nc_entry = NULL;
+    uint8_t sl2a_len = 0;
+    uint8_t *sl2a = (uint8_t *)(sl2a_opt + 1);
+
+    if ((sl2a_opt->len == 0) || ng_ipv6_addr_is_unspecified(&ipv6->src)) {
+        DEBUG("ndp: invalid source link-layer address option received\n");
+        return false;
+    }
+
+    while (pkt) {
+        if (pkt->type == NG_NETTYPE_NETIF) {
+            ng_netif_hdr_t *hdr = pkt->data;
+            sl2a_len = hdr->src_l2addr_len;
+            break;
+        }
+        pkt = pkt->next;
+    }
+
+    if (sl2a_len == 0) {  /* in case there was no source address
+                                 * (e.g. slip) */
+        sl2a_len = (sl2a_opt->len / 8) - sizeof(ng_ndp_opt_t);
+
+        /* ignore all zeroes at the end for length */
+        for (; sl2a[sl2a_len - 1] != 0x00; sl2a_len--);
+    }
+
+    switch (icmpv6_type) {
+#ifdef MODULE_NG_IPV6_ROUTER
+        case NG_ICMPV6_RTR_SOL:
+            nc_entry = ng_ipv6_nc_get(iface, &ipv6->src);
+
+            if (nc_entry != NULL) {
+                if ((sl2a_len != nc_entry->l2_addr_len) ||
+                    (memcmp(sl2a, nc_entry->l2_addr, sl2a_len) != 0)) {
+                    /* if entry exists but l2 address differs: set */
+                    nc_entry->l2_addr_len = sl2a_len;
+                    memcpy(nc_entry->l2_addr, sl2a, sl2a_len);
+
+                    /* set router to false. XXX: this will be different for
+                     * 6LoWPAN interfaces. */
+                    nc_entry->flags &= ~NG_IPV6_NC_IS_ROUTER;
+                    nc_entry->flags &= ~NG_IPV6_NC_STATE_MASK;
+                    nc_entry->flags |= NG_IPV6_NC_STATE_STALE;
+                }
+            }
+            else {
+                ng_ipv6_nc_add(iface, &ipv6->src, sl2a, sl2a_len,
+                               NG_IPV6_NC_STATE_STALE);
+
+            }
+
+            return true;
+#endif
+        case NG_ICMPV6_RTR_ADV:
+            /* entry was created as part of default router list so no
+             * check necessary */
+            nc_entry = ng_ipv6_nc_get(iface, &ipv6->src);
+
+            if ((sl2a_len == nc_entry->l2_addr_len) &&
+                (memcmp(sl2a, nc_entry->l2_addr, sl2a_len) == 0)) {
+                /* no update needed */
+                return true;
+            }
+
+            nc_entry->l2_addr_len = sl2a_len;
+            memcpy(nc_entry->l2_addr, sl2a, sl2a_len);
+
+            /* NG_IPV6_NC_IS_ROUTER flag already set (or unset if adv::ltime was 0) */
+            nc_entry->flags &= ~NG_IPV6_NC_STATE_MASK;
+            nc_entry->flags |= NG_IPV6_NC_STATE_STALE;
+
+            return true;
+
+        default:    /* wrong encapsulating message: silently discard */
+            DEBUG("ndp: silently discard sl2a_opt for ICMPv6 message type %"
+                  PRIu8 "\n", icmpv6_type);
+            return true;
+    }
+}
+
+static bool _handle_pi_opt(kernel_pid_t iface, ng_pktsnip_t *pkt,
```

Comment:
Because this PR is still WIP :-) It handles the PI (prefix information) option, which disseminate a prefix through a sub-net.